### PR TITLE
[BIOMAGE-1684] - Add DE warning

### PR DESCRIPTION
**Pull request by aerlaut**
# Description
#### Background
Running between-group DE comparison requires at least 2 samples to run. For comprehensive statistitics, we need the cells involved in the comparison to come from at least 3 different samples.

This PR provides a warning to the user to let them know if the group they have chosen to compare does not contain enough cells for complete statistical analysis.

#### To test
To test the check that is implemented here, you need to have a project with at least 4 samples.

- Go to Data Exploration, open the Differential Expression tab.
- Select comparison. between groups
- Choose a cluster located somewhere in the early middle of the cluster list (Cluster 0).
- Select one of the samples for the first group to compare
- Select "all" or "rest of samples".
- Change the cluster that is used as the basis.  You might be able to trigger the warning and the error alert.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1684

#### Link to staging deployment URL (or set N/A)
https://ui-agi-650.scp-staging.biomage.net

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [X] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [X] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [X] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [X] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [X] Relevant Github READMEs updated **or** no GitHub README updates required.
- [X] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx [hunk 1]
```diff
@@ -351,12 +362,31 @@ const DiffExprCompute = (props) => {
...
               showIcon
-            /> : <></>
+            /> :
+            isFormValid && canRunDiffExpr() === canRunDiffExprResults.WARNING ?
+              <Alert
+                message="Warning"
+                description={
+                  <>
+                    In the selected samples/groups, less than 3 samples contain enough cells to be compared.
+                    The differential expression analysis can be computed, but the adjusted p-value can not be calculated.
```
Comment (alexvpickering) [inline src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]:
```suggestion
                    For the selected comparison, there are fewer than 3 samples with the recommended minimum number of cells (10). Only logFC values will be calculated and results should be used for exploratory purposes only.
```

will need to adjust the test as well if you agree

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx [hunk 2]
```diff
@@ -150,13 +159,22 @@ const DiffExprCompute = (props) => {
...
+    const numCellSetSampleWithEnoughCells = numSampleWithEnoughCells(filteredCellSetCellIds)
+    const numCompareWithSampleWithEnoughCells = numSampleWithEnoughCells(filteredCompareWithCellIds)
+
+
+    if (numCellSetSampleWithEnoughCells === 0 || numCompareWithSampleWithEnoughCells === 0) return canRunDiffExprResults.ERROR;
+
+    const sumComparedSamples = numCellSetSampleWithEnoughCells + numCompareWithSampleWithEnoughCells;
+
+    if (sumComparedSamples < MIN_NUM_SAMPLES_SHOW_WARNING) return canRunDiffExprResults.WARNING;
+    if (sumComparedSamples < MIN_NUM_SAMPLES_SHOW_ERROR) return canRunDiffExprResults.ERROR;
```
Comment (alexvpickering) [inline src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]:
```suggestion
    if (sumComparedSamples < MIN_NUM_SAMPLES_SHOW_ERROR) return canRunDiffExprResults.ERROR;
    if (sumComparedSamples < MIN_NUM_SAMPLES_SHOW_WARNING) return canRunDiffExprResults.WARNING;
```
These are in the wrong order. The second check will never evaluate to true.

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx [hunk 3]
```diff
@@ -22,6 +22,15 @@ const getCellSetKey = (name) => (name?.split('/')[1] || name);
 const getRootKey = (name) => name?.split('/')[0];
 
 const MIN_NUM_CELLS_IN_GROUP = 10;
+const MIN_NUM_SAMPLES_SHOW_ERROR = 2;
+const MIN_NUM_SAMPLES_SHOW_WARNING = 3;
```
Comment (alexvpickering) [inline src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: I would call these e.g. `MIN_NUM_SAMPLES_OMIT_ERROR` or similar. Alternatively, decrement these by one and make the checks below `<=`.

I don't think it makes lexical sense right now as you only show the error if there are less than (but not equal to) the minimum number of samples to show an error (`MIN_NUM_SAMPLES_SHOW_ERROR`).

**Review by alexvpickering — CHANGES_REQUESTED** [review]
I added a couple of comments/suggestions but overall looks top notch as always. Let me know when you address them and I'll approve

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: Righ! Thanks for catching it Alex!

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 1 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: No problme, added :+1:

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 3 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: Hmm.. yes, I think it's easier to understand.

**Review by alexvpickering — COMMENTED** [review]
The changes you made look great Agi! I'm feeling off today and didn't carefully check the code so you may want to have an engineer take a look as well

**Review by alexvpickering — APPROVED** [review]

File: src/__test__/utils/cellSets/getAllCells.test.js [hunk 4]
```diff
@@ -25,7 +25,7 @@ describe('Get all cells', () => {
 
     expect(allCellIds[0]).toMatchInlineSnapshot(`
       Object {
-        "cellId": 11,
+        "cellId": 1,
```
Comment (cosa65) [inline src/__test__/utils/cellSets/getAllCells.test.js:28]: Why was this changed?

File: src/__test__/utils/plotSpecs/generateEmbeddingCategoricalSpec.test.js [hunk 5]
```diff
@@ -19,31 +19,18 @@ describe('filterCells', () => {
...
-      .find(({ key }) => key === 'louvain').children
-      .map(({ key, name, color }) => ({ key, name, color }));
-
-    // Contains all the louvain cellSets (and in the correct order)
-    expect(result.cellSetLegendsData).toEqual(expectedCellSetKeys);
-
-    // CellIds are filtered fine
-    expect(result).toMatchSnapshot();
-  });
-
```
Comment (cosa65) [inline src/__test__/utils/plotSpecs/generateEmbeddingCategoricalSpec.test.js:35]: Why was this test removed?

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx [hunk 6]
```diff
@@ -150,13 +159,21 @@ const DiffExprCompute = (props) => {
           numCellsPerSampleInCellSet[sampleIdx] += 1;
         });
 
-      return numCellsPerSampleInCellSet.find(numCells => numCells > MIN_NUM_CELLS_IN_GROUP)
+      return numCellsPerSampleInCellSet.filter(numCells => numCells >= MIN_NUM_CELLS_IN_GROUP).length;
     }
 
-    const cellSetHasSampleWithEnoughCells = hasSampleWithEnoughCells(filteredCellSetCellIds)
-    const compareWithHasSampleWithEnoughCells = hasSampleWithEnoughCells(filteredCompareWithCellIds)
+    const numCellSetSampleWithEnoughCells = numSampleWithEnoughCells(filteredCellSetCellIds)
```
Comment (cosa65) [inline src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: `canRunDiffExpr` seems pretty long and complex, do you think it could be extracted to a function an imported here? And if there's also other functions like this maybe they should be changed too.

The jsx file could just have the logic more directly related to what to render and call functions for the processing, etc.

File: src/__test__/utils/plotSpecs/generateEmbeddingCategoricalSpec.test.js (hunk 5 above)
Comment (aerlaut) [reply src/__test__/utils/plotSpecs/generateEmbeddingCategoricalSpec.test.js:35]: This is a duplicated test with the one on line 111 in the old file

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 6 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: This logic is only used in this file. It is also tightly coupled with the variables and state of the function, so I think there's no need to extract it out.

Some of the cell set operations inside the function is repeated a lot throughout the code base though. For these operations, I agree we can factor it out into separate functions. However, I think it is a general refactor better left for another ticket / hackathon.

File: src/__test__/utils/cellSets/getAllCells.test.js (hunk 4 above)
Comment (aerlaut) [reply src/__test__/utils/cellSets/getAllCells.test.js:28]: I added cells and made changes to the mock `__test__/data/cell_sets.json` file to be able to test the check introduced here. If you look at the diff for [cell_sets.json](https://github.com/hms-dbmi-cellenics/ui/pull/650/files#diff-4a1bbbf10ce496597b0ecff72b7426bc5942ba042d604319475c5110650c41d1R145-R153), the first cell id of sample `"b62028a1-ffa0-4f10-823d-93c9ddb88898"` was initiallly `11`.  I rearranged the cells in the sample and it was changed to 1.

Comment (cosa65) [reply src/__test__/utils/cellSets/getAllCells.test.js:28]: Woah, I would have though that changing cell_sets.json would have caused the need for way more changes in unit tests since it is being used in virtually all of the new unit tests.

Next time, however, if you need a new dataset I think I would recommend just creating a different `.json` for your use case instead of changing an existing one and having to update a lot of different tests.

Comment (cosa65) [reply src/__test__/utils/cellSets/getAllCells.test.js:28]: I think sometihng good about the previous `cell_sets.json` was that it was simple so it was easy to understand the tests we were making with it

Comment (aerlaut) [reply src/__test__/utils/cellSets/getAllCells.test.js:28]: this is also simple - there's just more cells

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 6 above)
Comment (cosa65) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: I agree, we don't need to do a general refactor on the cell set operations here.

My reason for separating responsibilities was not mainly about reusability, more about readability and being able to have a clearer separation, an example of something similar could be `processUpload` or `calculateVolcanoDataPoints`, they are all about doing more involved calculations and they are extracted into functions even when we could easily just have them i one place in the `jsx` that uses them.

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx [hunk 7]
```diff
@@ -14,14 +14,20 @@ import { loadCellSets } from 'redux/actions/cellSets';
...
 const getCellSetKey = (name) => (name?.split('/')[1] || name);
 const getRootKey = (name) => name?.split('/')[0];
 
-const MIN_NUM_CELLS_IN_GROUP = 10;
+const canRunDiffExprResults = {
+  TRUE: 'TRUE',
+  FALSE: 'FALSE',
+  INSUFFICIENT_CELLS_WARNING: 'INSUFFICIENT_CELLS_WARNING',
+  INSUFFCIENT_CELLS_ERROR: 'INSUFFCIENT_CELLS_ERROR'
+}
```
Comment (cosa65) [inline src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: We could move this to checkCanRunDiffExpr

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx [hunk 8]
```diff
@@ -97,66 +103,23 @@ const DiffExprCompute = (props) => {
...
-        });
-
-      return numCellsPerSampleInCellSet.find(numCells => numCells > MIN_NUM_CELLS_IN_GROUP)
-    }
+  const canRunDiffExpr = useCallback(() => {
 
-    const cellSetHasSampleWithEnoughCells = hasSampleWithEnoughCells(filteredCellSetCellIds)
-    const compareWithHasSampleWithEnoughCells = hasSampleWithEnoughCells(filteredCompareWithCellIds)
+    if (selectedComparison === ComparisonType.WITHIN) return canRunDiffExprResults.TRUE;
+    if (!basis || !cellSet || !compareWith || !cellIdToSampleMap.length > 0) { return canRunDiffExprResults.FALSE; }
```
Comment (cosa65) [inline src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: We could also move this to checkCanRunDiffExpr

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 7 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: moved

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 8 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: moved

**Review by cosa65 — APPROVED** [review]

**Comment by cosa65** [comment]
aprobado
